# Deriving New Variables

## Content
#### 1. Import libraries and data
#### 2. Create new column "price_range"
#### 3. Create new column "busiest day"
#### 4. Create new column "busiest days"
#### 5. Create new column 'busiest_period_of_day'
#### 6. Export data

# 1. Import libraries and data

In [58]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [59]:
# Create project folder path
path = r'C:\Users\Lara\Career Foundry Projects\21-09-2023 Instacart Basket Analysis'

In [60]:
# Import dataset orders_products_merged.pkl
df_ords_prods_merged = pd.read_pickle (os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [61]:
df_ords_prods_merged.shape

(32434212, 13)

In [62]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices  
0         Soda        77              7     9.0  
1         Soda        77              7     9.0  
2         Soda        77              7     9.0  
3         Soda        77              7     9.0  
4         Soda        77              7     9.0

# 2. Create new column "price_range"

#### Use loc() function to apply the 3 conditions, each in separate cell

In [63]:
# Define conditions

df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range'] = 'High-range product'

In [64]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range'] = 'Mid-range product' 

In [65]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range'] = 'Low-range product'

In [66]:
# Count all values
df_ords_prods_merged['price_range'].value_counts(dropna = False)

price_range
Mid-range product     21890146
Low-range product     10126384
High-range product      417682
Name: count, dtype: int64

In [67]:
# Check current shape and head of dataframe
df_ords_prods_merged.shape

(32434212, 14)

In [68]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_range  
0         Soda        77              7     9.0  Mid-range product  
1         Soda        77              7     9.0  Mid-range product  
2         Soda        77              7     9.0  Mid-range product  
3         Soda        77              7     9.0  Mid-range product  
4         Soda        77              7     9.0  Mid-range product

# 3. Create new column "busiest day"

#### Create column with values implying how busy each day of the week was.
#### Run a for-loop for every row and assign desired value to one busiest day, one slowest day and to other 5 days asign regularly busy.

In [69]:
# Count all orders for each day
df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6209632
1    5665830
6    4500246
2    4217766
5    4209449
3    3844096
4    3787193
Name: count, dtype: int64

#### Busiest day is 0 = Saturday, least busy is 4 = Wednesday and the rest are regularly busy.

In [70]:
# Create an empty list first for for-loop

result = []

# for-loop for level of busyness of days of the week

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [71]:
# Create new column in dataframe
df_ords_prods_merged['busiest_day'] = result

In [72]:
# Count all values of new column
df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22437387
Busiest day        6209632
Least busy         3787193
Name: count, dtype: int64

In [73]:
# Check current shape and head of dataframe
df_ords_prods_merged.shape

(32434212, 15)

In [74]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day  
0  Regularly busy  
1  Regularly busy  
2  Regularly busy  
3      Least busy  
4      Least busy

# 4. Create new column "busiest days"

#### First drop column "busiest day" - it is no longer needed.
#### Run a for-loop for every row and assign desired value to 2 busiest days, 2 slowest days and to other 3 days asign regularly busy.
#### Busiest days are 0 = Saturday and 1 = Sunday, slowest days are 4 = Wednesday and 3 = Tuesday and the rest are Regularly busy

In [75]:
# Drop column 'busiest_day' and overwrite dataframe
df_ords_prods_merged = df_ords_prods_merged.drop(columns = ['busiest_day'])

In [76]:
# Check if column was dropped successfully
df_ords_prods_merged.shape

(32434212, 14)

In [77]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_range  
0         Soda        77              7     9.0  Mid-range product  
1         Soda        77              7     9.0  Mid-range product  
2         Soda        77              7     9.0  Mid-range product  
3         Soda        77              7     9.0  Mid-range product  
4         Soda        77              7     9.0  Mid-range product

In [78]:
# Create an empty list first for for-loop

result = []

# for-loop for level of busyness of days of the week

for value in df_ords_prods_merged["orders_day_of_week"]:
  if (value == 0 or value == 1):
    result.append("Busiest days")
  elif (value == 4 or value == 3):
    result.append("Slowest days")
  else:
    result.append("Regularly busy")

In [79]:
# Create new column in dataframe
df_ords_prods_merged['busiest_days'] = result

In [81]:
# Count all values of new column
df_ords_prods_merged['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12927461
Busiest days      11875462
Slowest days       7631289
Name: count, dtype: int64

#### This numbers correspond to sums of orders for days that are of particular level of busyness, for example Busiest days = 11875462 = 6209632 + 5665830 which represent number of orders on days 0 = Saturday and 1 = Sunday. 

In [82]:
# Check current shape and head of dataframe
df_ords_prods_merged.shape

(32434212, 15)

In [83]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

     busiest_days  
0  Regularly busy  
1    Slowest days  
2    Slowest days  
3    Slowest days  
4    Slowest days

# 5. Create new column 'busiest_period_of_day'

In [84]:
# Count all orders for each hour of the day
df_ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2764390
11    2738585
14    2691563
15    2664522
13    2663272
12    2620800
16    2537469
9     2456661
17    2089452
8     1719952
18    1637922
19    1259382
20     977017
7      891928
21     796362
22     634737
23     402612
6      290770
0      218942
1      115786
5       88057
2       69431
4       53283
3       51317
Name: count, dtype: int64

#### From this numbers we can see that most orders happened between hours 9 and 16 with each hour, having between around 2.4 M and 2.8 M orders. Least number of orders happened between hours 0 and 6 and at 23 with number of orders ranging from around 50 K to 400 K. At all other hours we have awerage number of order between about 600 K and 2.1 M. K=1000, M=1000000
#### I chose to group hours of days like this into 3 periods, because number of orders were similar in each group and I was looking for somewhat significant drop or rise between two hours in list above to determine when will period end or begin. For example, for hours 23 and 6 I noticed that for hours 21, 22 and 7 difference between number of orders is about 100 K, but between 23 and 6 it is 230 K and on the other side for 23, 6 and 0 difference in number of order is about 100K. 
#### This method actually produced 3 period of a day with equal lenght of 8 hours.

In [85]:
# Create an empty list first for for-loop

result = []

# for-loop for level of busyness of periods in a day

for value in df_ords_prods_merged["order_hour_of_day"]:
  if (9 <= value <= 16):
    result.append("Most orders")
  elif (0 <= value <= 6 or value == 23):
    result.append("Fewest orders")
  else:
    result.append("Awerage orders")

In [86]:
# Create new column in dataframe
df_ords_prods_merged['busiest_period_of_day'] = result

In [87]:
# Count all values of new column
df_ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21137262
Awerage orders    10006752
Fewest orders      1290198
Name: count, dtype: int64

#### Numbers correspond to sums of number of order for each hour of the day that belongs to certain period. For example for label 'Fewest order' 1290198 = 218942 + 115786 + 69431 + 51317 + 53283 + 88057 + 290770 + 402612 which is a sum of number of order for hours 0,1, 2, 3, 4, 5, 6 and 23

In [88]:
# Check current shape and head of dataframe
df_ords_prods_merged.shape

(32434212, 16)

In [89]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices        price_range  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

     busiest_days busiest_period_of_day  
0  Regularly busy        Awerage orders  
1    Slowest days        Awerage orders  
2    Slowest days           Most orders  
3    Slowest days        Awerage orders  
4    Slowest days           Most orders

# 6. Export data

In [90]:
# Export dataframe as pikle format
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_new_cols.pkl'))